File: Tree/Tree_Breadth_First/BFSTree.java
import java.util.Queue;
import java.util.LinkedList;

/**
 * Breadth-First Binary Search Tree implementation class.
 * @author Yasin Radi
 */
public class BFSTree {

    /** Tree Node implementation. */
    private class Node {
        // Data held by the node
        private int data;

        // Left child of node
        private Node left;

        // Right child of node
        private Node right;

        private Node(int data) {
            this.data = data;
            this.right = null;
            this.left = null;
        };
    };

    // Root of the Tree
    private Node root;

    // Constructor
    public BFSTree() {
        this.root = null;
    };

    /**
     * Helper that performs a std BST insert.
     * @param root  Node
     * @param node  Node
     * @return      Node
     */
    private Node insertHelper(Node root, Node node) {
        // If tree is empty, return new node
        if (root == null) {
            root = node;
            return node;
        }

        // Otherwise, recur down the tree
        if (node.data < root.data) {
            root.left = insertHelper(root.left, node);
        } else if (node.data > root.data) {
            root.right = insertHelper(root.right, node);
        }

        // Return the unchanged node
        return root;
    };

    /**
     * Recursive helper function for inOrder traversal.
     * @param node  Node
     */
    private void inOrderHelper(final Node node) {
        if (node == null) {
            return;
        }

        inOrderHelper(node.left);
        System.out.print(node.data + " ");
        inOrderHelper(node.right);
    };

    /**
     * Helper function for level order traversal.
     * @param node  Node
     */
    private void levelOrderHelper(final Node node) {
        if (node == null) {
            return;
        }

        Queue<Node> q = new LinkedList<>();
        q.add(node);

        while (!q.isEmpty()) {
            Node tmpNode = q.poll();
            System.out.print(tmpNode.data + " ");
            
            if (tmpNode.left != null) {
                q.add(tmpNode.left);
            }

            if (tmpNode.right != null) {
                q.add(tmpNode.right);
            }
        }
    };

    /**
     * Inserts a node with the given data into the tree.
     * @param data  int
     */
    public void insert(final int data) {
        // Create node to be inserted
        Node node = new Node(data);

        // Perform a normal BST insert.
        this.root = insertHelper(this.root, node);
    };

    /** InOrder traversal tree content display. */
    public void inOrder() {
        inOrderHelper(this.root);
        System.out.println();
    };

    /** Level order traversal tree content display. */
    public void levelOrder() {
        levelOrderHelper(this.root);
        System.out.println();
    };
};

File: Tree/Tree_Breadth_First/bfs_tree.py
from collections import deque

class Node:
    """ Tree Node implementation. 
    
    Attributes:
    -----------

    left : Node
        Left child of the node.
    
    right : Node
        Right child of the node.
    
    Parameters:
    -----------
    data : int
        Data to be held by the node.
    
    """
    def __init__(self, data):
        self.data = data
        self.left : Node = None
        self.right : Node = None

class BFSTree:
    """ Breadth-First Binary Search Tree implementation class. 
    
    Attributes:
    -----------
    root : Node
        Root node of the tree.

    """
    def __init__(self):
        self.root : Node = None
    
    def _insertHelper(self, root, node):
        """ Performs a STD BST insert. 
        
        Parameters:
        ----------
        root : Node
            Root of the tree.
        
        node : Node
            Node to be inserted.
        
        Returns:
        --------
        The root of the tree or the same node if tree is empty. 
        
        """
        # If tree is empty, return new node
        if root is None:
            root = node
            return node
        
        # Otherwise, recurr down the tree
        if node.data < root.data:
            root.left = self._insertHelper(root.left, node)
        elif node.data > root.data:
            root.right = self._insertHelper(root.right, node)
        
        # Return root of the tree
        return root

    def _inorderHelper(self, node):
        """ Recursive helper function for inorder traversal.

        Parameters:
        -----------
        node : Node
            Node whose content will be displayed. 
            
        """
        if node is None:
            return
        
        self._inorderHelper(node.left)
        print(node.data, end=' ')
        self._inorderHelper(node.right)

    def _levelOrderHelper(self, node):
        """ Helper function for level order traversal.

        Parameters:
        -----------
        node : Node
            Starting node for traversal content display. 
            
        """
        if node is None:
            return
        
        q = deque()
        q.append(node)

        while len(q) > 0:
            tmpNode = q.popleft()
            print(tmpNode.data, end=' ')

            if tmpNode.left is not None:
                q.append(tmpNode.left)
            
            if tmpNode.right is not None:
                q.append(tmpNode.right)

    def insert(self, data):
        """ Inserts a new node into the tree.
        
        Parameters:
        -----------
        data : int
            Data to be inserted into the tree.
        
        """
        # Create node to be inserted.
        node = Node(data)

        # Perform a BST insert.
        # Update root if apliable
        self.root = self._insertHelper(self.root, node)
        
    def inorder(self):
        """ Inorder traversal tree content display. """
        self._inorderHelper(self.root)
        print()
    
    def levelOrder(self):
        """ Level Order traversal tree content display. """
        self._levelOrderHelper(self.root)
        print()